core/auth/src/main/java/software/amazon/awssdk/auth/credentials/WebIdentityTokenCredentialsProvider.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.credentials;
+
+import static software.amazon.awssdk.utils.StringUtils.trim;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.auth.credentials.internal.WebIdentityCredentialsUtils;
+import software.amazon.awssdk.auth.credentials.internal.WebIdentityTokenCredentialProperties;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.utils.ToString;
+
+/**
+ * A credential provider that will read web identity token file path, aws role arn
+ * and aws session name from system properties or environment variables for using
+ * web identity token credentials with STS. Use of this credentials provider requires
+ * the 'sts' module to be on the classpath.
+ */
+@SdkPublicApi
+public class WebIdentityTokenCredentialsProvider implements AwsCredentialsProvider {
+
+    private final AwsCredentialsProvider credentialsProvider;
+    private final RuntimeException loadException;
+
+    private WebIdentityTokenCredentialsProvider(BuilderImpl builder) {
+        AwsCredentialsProvider credentialsProvider = null;
+        RuntimeException loadException = null;
+
+        try {
+            Path webIdentityTokenFile;
+            String roleArn;
+            Optional<String> roleSessionName;
+
+            // Don't allow a mix of environment variables and builder configured values
+            if (builder.webIdentityTokenFile != null) {
```

Comment:
I'm a little worried that builder.webIdentityTokenFile = null means "use environment variables" even if they specified the roleArn and roleSessionName.

Do we have a precedent for this?